### PR TITLE
[SR-9425][Sema] Use derived conformance as witness when possible for == operator of raw representable enums

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
There are two overloads involved in this case 
* `<Self where Self : Equatable> (Self.Type) -> (Self, Self) -> Bool`
* `<T where T : RawRepresentable, T.RawValue : Equatable> (T, T) -> Bool`

Solver is finding the first one as best because the ranking rule of non-generic is better than generic.
However in CSApply https://github.com/apple/swift/blob/22506f9bdb009c0c82410406d825c87c08af7f29/lib/Sema/CSApply.cpp#L621
currently it is finding the `<T where T : RawRepresentable, T.RawValue : Equatable> (T, T) -> Bool` as the `Equatable` (`WitnessChecker::findBestWitness` find only this one)witness which ends up applied. What given the implementation of it:
```swift
@inlinable // trivial-implementation
public func == <T: RawRepresentable>(lhs: T, rhs: T) -> Bool
  where T.RawValue: Equatable {
  return lhs.rawValue == rhs.rawValue
}
```
When raw value is a `String` the codegen ends up with a string compare. 

This PR adjusts `ConformanceChecker::resolveWitnessViaLookup` to not consider this witness as best match for swift raw representable enums when derive `Equatable` is possible(favor derived). Therefore avoiding a str compare for those cases. 

**Note**
More details on the JIRA issue
Also, this PR although fixes the issue and makes sense to me, I'm not familiar enough with this to know if it is really the best fix. 

cc @eeckstein @xedin 

Resolves SR-9425, rdar://46555205

### CODE REVIEW

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -1324,6 +1324,19 @@ bool WitnessChecker::findBestWitness(
     anyFromUnconstrainedExtension = false;
     bestIdx = 0;
 
+    auto &ctx = getASTContext();
+    bool isEquatableConformance =
+        conformance && conformance->getProtocol() ==
+                           ctx.getProtocol(KnownProtocolKind::Equatable);
+
+    bool isSwiftRawRepresentableEnum = false;
+    if (conformance && conformance->getType()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp:1333]: conformance->getType() is never going to be null

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -1324,6 +1324,19 @@ bool WitnessChecker::findBestWitness(
...
+      auto enumType = conformance->getType()->getAs<EnumType>();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp:1334]: This won't work if the enum is generic. You should use conformance->getDeclContext()->getSelfNominalDecl() instead

File: lib/Sema/TypeCheckProtocol.cpp [hunk 3]
```diff
@@ -1336,6 +1349,29 @@ bool WitnessChecker::findBestWitness(
       auto match = matchWitness(ReqEnvironmentCache, Proto, conformance,
                                 DC, requirement, witness);
       if (match.isViable()) {
+        if (canDerive && isSwiftRawRepresentableEnum &&
+            isEquatableConformance) {
+          // For enum types that can derive equatable conformance, let's always
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp:1354]: This seems like an overly specific check to add to findBestWitness(). We wouldn't want multiple bits of code like this to accumulate here over time. I would prefer it if we could find another solution here.

Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckProtocol.cpp:1354]: Thanks for the review @slavapestov! 

>  We wouldn't want multiple bits of code like this to accumulate here over time. I would prefer it if we could find another solution here.

That makes sense, I'll try to come up with a different approach :)

Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckProtocol.cpp]: Made some adjustments in the approach, let me know what you think

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

**Comment by LucianoPAlmeida** [comment]
There is a benchmark for this https://github.com/apple/swift/blob/main/benchmark/single-source/OpenClose.swift
Just curious about the impact :)

@swift-ci Please smoke benchmark

File: lib/Sema/TypeCheckProtocol.cpp [hunk 4]
```diff
@@ -3832,6 +3832,25 @@ void ConformanceChecker::checkNonFinalClassWitness(ValueDecl *requirement,
...
 }
 
+// If the given witness matches a generic RawRepresentable function conforming
+// with a given protocol e.g. `func == <T : RawRepresentable>(lhs: T, rhs: T) ->
+// Bool where T.RawValue : Equatable`
+static bool isRawRepresentableGenericFunction(
+    ASTContext &ctx, const ValueDecl *witness,
+    const NormalProtocolConformance *conformance) {
+  auto *fnDecl = dyn_cast<AbstractFunctionDecl>(witness);
+  return fnDecl && fnDecl->isGeneric() &&
```
Comment (CodaFi) [inline lib/Sema/TypeCheckProtocol.cpp:3842]: Let's try matching more aggressively. We can ensure `isStdlibDecl` and that there aren't further requirements. This'll let users that have a global == witness of their own continue to compile against it.

**Comment by milseman** [comment]
This has been such a long-standing perf issue. Thanks for fixing this!

File: lib/Sema/TypeCheckProtocol.cpp (hunk 4 above)
Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckProtocol.cpp]: Make sense just fixed, thanks @CodaFi

**Comment by CodaFi** [comment]
@swift-ci test

**Review by CodaFi — APPROVED** [review]

**Pull request merged**